op-node: Add dedicated PlasmaConfig struct to rollup.Config

trianglesphere opened the pull request:
**Description**

This commit adds a dedicated PlasmaConfig section to the rollup config. This collects all Plasma Mode configuration to the same place and enables future expansion of Plasma Mode configuration to be confined to a single location.

In the interim, the op-node can read from both the legacy fields and the dedicated struct. If both are set, they must be consistent. The genesis creation tooling currently only writes with the new config.

This also updates to a new version of the superchain registry to enable Plasma Mode chains in the superchain registry to properly set the rollup config when configured by the superchain flag name.

**Tests**

Tested locally & the output JSON looks good. Unfortunately `omitempty` does not work on the address type.


**Metadata**

- Fixes https://github.com/ethereum-optimism/optimism/issues/10472

bitwiseguy commented on op-node/rollup/types.go:
```diff
@@ -50,6 +50,17 @@ type Genesis struct {
 
+type PlasmaConfig struct {
+	// L1 DataAvailabilityChallenge contract proxy address
+	DAChallengeAddress common.Address `json:"da_challenge_contract_address,omitempty"`
```
Did you try changing the type to a pointer to get the `omitempty` to work correctly ?
```
DAChallengeAddress *common.Address `json:"da_challenge_contract_address,omitempty"`
```

trianglesphere replied:
No, but I'm not super worried about `omitempty` here

bitwiseguy commented on op-node/rollup/superchain.go:
```diff
@@ -48,6 +48,9 @@ func LoadOPStackRollupConfig(chainID uint64) (*Config, error) {
 		return nil, fmt.Errorf("unable to retrieve deposit contract address")
 	}
 
+	// var plasma *PlasmaConfig
+	// if Config.chConfig.
+
```
Seems like this comment could be removed

ajsutton reviewed (commented):
CI apparently said no:

```
testing.tRunner.func1.2({0x1dfd2a0, 0x40b1d60})
        /usr/local/go/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1548 +0x397
panic({0x1dfd2a0?, 0x40b1d60?})
        /usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/ethereum-optimism/optimism/op-node/rollup.(*Config).GetOPPlasmaConfig(0xc0034b6050)
        /root/project/op-node/rollup/types.go:452 +0x3c
github.com/ethereum-optimism/optimism/op-node/node.(*OpNode).initL2(0xc000324ea0, {0x306d560, 0x45f7280}, 0xc0034b6000, {0x30812b0, 0xc002206190})
        /root/project/op-node/node/node.go:390 +0x33f
github.com/ethereum-optimism/optimism/op-node/node.(*OpNode).init(0xc000324ea0, {0x306d560, 0x45f7280}, 0xc0034b6000, {0x30812b0, 0xc002206190})
        /root/project/op-node/node/node.go:131 +0x1de
github.com/ethereum-optimism/optimism/op-node/node.New({0x306d560, 0x45f7280}, 0xc0034b6000, {0x3081418, 0xc001e08630}, {0x30812b0, 0xc002206190}, {0x0, 0x0}, 0xc00340ef00)
```

trianglesphere committed:
op-node: Add dedicated PlasmaConfig struct to rollup.Config

This commit adds a dedicated PlasmaConfig section to the rollup
config. This collects all Plasma Mode configuration to the same place
and enables future expansion of Plasma Mode configuration to be confined
to a single location.

In the interim, the op-node can read from both the legacy fields and the
dedicated struct. If both are set, they must be consistent. The genesis
creation tooling currently only writes with the new config.

trianglesphere committed:
Set plasma mode from superchain registry

trianglesphere commented:
I should have fixed CI. This also now sets the local plasma config from the superchain plasma config

trianglesphere committed:
Remove LegacyUsePlasma usage

ajsutton commented on op-node/rollup/types.go:
```diff
@@ -306,6 +323,31 @@ func (cfg *Config) Check() error {
+// If the legacy values are set, they are copied to the new location. If both are set, they are check for consistency.
+func validatePlasmaConfig(cfg *Config) error {
+	if cfg.LegacyUsePlasma && cfg.PlasmaConfig == nil {
+		// copy from top level to plasma config
+		cfg.PlasmaConfig = &PlasmaConfig{
+			DAChallengeAddress: cfg.LegacyDAChallengeAddress,
+			DAChallengeWindow:  cfg.LegacyDAChallengeWindow,
+			DAResolveWindow:    cfg.LegacyDAResolveWindow,
+		}
```
Should we nil out the legacy values so that if this config is written out again they will then be omitted? Not sure it matters at this point in the code though.

trianglesphere replied:
I don't think we go through a read/write cycle. The two writes are from the genesis creation tool  (values from deploy config) and the network config dump tool (network config dump requires the flag)

ajsutton reviewed (approved):
LGTM.
